src/book.js

```diff
@@ -653,78 +677,104 @@ export class Book {
+            let ygap = finaly -  pagey * sy;
+            return padding['top'] + ygap/2 ;
+        }
+        let yBottomShiftFunc = function() {
+            let ygap = finaly -  pagey * sy;
+            return padding['bottom'] + ygap/2 ;
+        }
+        let xPdfWidthFunc = function() {
+            return pagex * sx - padding['fore_edge'] - padding['binding'];
+        }
+        let yPdfHeightFunc = function() {
+            return pagey * sy - padding['top'] - padding['bottom'];
+        }
+        return {
+            layout: layout,
+            rawPdfSize: [this.cropbox.width, this.cropbox.height],
+            pdfScale: [sx, sy],
+            pdfSize: [pagex, pagey],
+            layoutCell: [finalx, finaly],
+            padding: padding,
+
+            xForeEdgeShiftFunc: xForeEdgeShiftFunc,
+            xBindingShiftFunc: xBindingShiftFunc,
+            xPdfWidthFunc: xPdfWidthFunc,
+            yPdfHeightFunc: yPdfHeightFunc,
+            yTopShiftFunc: yTopShiftFunc,
+            yBottomShiftFunc: yBottomShiftFunc,
+
+            positioning: positioning
+        }
+    }
 
-                // value tracking white space to left and right of an open book/page spread
-                let leftRightPageGap = ([90,-90].includes(col)) ? ypad : xpad;
+    /**
+     * When considering page size, don't forget to take into account 
+     *  this.padding_pt's ['top','bottom','binding','fore_edge'] values
+     *
+     * @return an array of objects in the form {rotation: col, sx: sx, sy: sy, x: x, y: y}
+     */
+    calculatelayout(alt_folio){
+        // vampire
+        let l = this.calculate_dimensions()
+        let cellWidth = l.layoutCell[0]
+        let cellHeight = l.layoutCell[1]
+        let positions = []
 
-                // amount to inset by, relative to fore edge, on left side of book
-                let xForeEdgeShift = padding['fore_edge'] + ((this.page_positioning == 'centered' ) ? leftRightPageGap : 2 * leftRightPageGap);
-                // amount to inset by, relative to binding, on right side of book
-                let xBindingShift = padding['binding'] + ((this.page_positioning == 'centered' ) ? leftRightPageGap : 0);
+        l.layout.rotations.forEach((row, i) => {
+            row.forEach((col, j) => {
+                let xForeEdgeShift = l.xForeEdgeShiftFunc();
+                let xBindingShift = l.xBindingShiftFunc();
+                let yTopShift = l.yTopShiftFunc();
+                let yBottomShift = l.yBottomShiftFunc();
 
                 let isLeftPage = j % 2 == 0; //page on 'left' side of open book
-                let x = (j * finalx) + ((j % 2 == 0 ) ? xForeEdgeShift : xBindingShift);
-                let y = (i * finaly) + ypad + padding['bottom'] * sy;
+                let x = (j * cellWidth) + ((isLeftPage) ? xForeEdgeShift : xBindingShift);
+                let y = (i * cellHeight) + yBottomShift;
 
                 if ([-180].includes(col)) { // upside-down page
```

Comment:
Not changed in this, but I think this makes more sense as just an equality check? Rather than looking against a one-item list.